src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -341,170 +603,333 @@ function saturation_shum(T, ρ, q_liq=0, q_ice=0)
+    p_vs        = saturation_vapor_pressure(T, LH_0, Δcp)
 
     return saturation_shum_from_pressure(T, ρ, p_vs)
 
 end
 
+"""
+    saturation_shum(ts::ThermodynamicState)
 
+Compute the saturation specific humidity, given a thermodynamic state `ts`.
 """
-    saturation_shum_from_pressure(T, ρ, p_vs)
+saturation_shum(ts::ThermodynamicState) =
+  saturation_shum(air_temperature(ts), air_density(ts), phase_partitioning_eq(ts)...)
 
-Compute the saturation specific humidity, given the ambient air temperature `T`,
-density `ρ`, and the saturation vapor pressure `p_vs`.
 """
-function saturation_shum_from_pressure(T, ρ, p_vs)
+    saturation_shum_from_pressure(T, ρ, p_vs)
 
-    return min(typeof(ρ)(1), p_vs / (ρ * R_v * T))
+Compute the saturation specific humidity, given
 
-end
+ - `T` ambient air temperature,
+ - `ρ` density
+ - `p_vs` saturation vapor pressure
+"""
+saturation_shum_from_pressure(T::DT, ρ::DT, p_vs::DT) where {DT<:Real} =
+  min(1, p_vs / (ρ * DT(R_v) * T))
 
 """
     saturation_excess(T, ρ, q_tot, q_liq=0, q_ice=0)
 
-Compute the saturation excess in equilibrium, given the ambient air temperature
-`T`, the (moist-)air density `ρ`, the total specific humidity `q_tot`, and,
-optionally, the liquid specific humidity `q_liq`, and the ice specific humidity `q_ice`.
+The saturation excess in equilibrium where
+
+ - `T` temperature
+ - `ρ` (moist-)air density
+ - `q_tot` total specific humidity
+and, optionally,
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
 
 The saturation excess is the difference between the total specific humidity `q_tot`
 and the saturation specific humidity in equilibrium, and it is defined to be
 nonzero only if this difference is positive.
 """
-function saturation_excess(T, ρ, q_tot, q_liq=0, q_ice=0)
+saturation_excess(T::DT, ρ::DT, q_tot::DT, q_liq::DT=DT(0), q_ice::DT=DT(0)) where {DT} =
+  max(0, q_tot - saturation_shum(T, ρ, q_liq, q_ice))
 
-    return max(typeof(q_tot)(0), q_tot - saturation_shum(T, ρ, q_liq, q_ice))
+"""
+    saturation_excess(ts::ThermodynamicState)
 
-end
+Compute the saturation excess in equilibrium,
+given a thermodynamic state `ts`.
+"""
+saturation_excess(ts::ThermodynamicState) =
+  saturation_excess(air_temperature(ts), air_density(ts), ts.q_tot, phase_partitioning_eq(ts)...)
 
 """
     liquid_fraction(T[, q_liq=0, q_ice=0])
 
-Return the fraction of condensate that is liquid.
+The fraction of condensate that is liquid where
+
+ - `T` temperature
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
 
 If the optional input arguments `q_liq` and `q_ice` are not given or are zero, the
 fraction of liquid is a function that is 1 above `T_freeze` and goes to zero below
 `T_freeze`. If `q_liq` or `q_ice` are nonzero, the liquid fraction is computed from
 them.
 """
-function liquid_fraction(T, q_liq=0, q_ice=0)
+function liquid_fraction(T::DT, q_liq::DT=DT(0), q_ice::DT=DT(0)) where {DT<:Real}
```

Comment:
It would be good to have two functions `liquid_fraction_equil` and `liquid_fraction_nonequil`. `liquid_fraction_equil` should always be the Heaviside function. `liquid_fraction_nonequil` for now can be equal to `liquid_fraction_equil` but eventually can be a more general function.